docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+* Use `DispatchProxy` to create a proxy within the user ALC, pointing to an object in the target ALC. This also requires a shared assembly.
+
+Anything requiring a shared assembly is inadequate for what we're trying to do, as one of the main benefits of AppDomains having totally separate types was with version compatibility. In .NET Framework, different versions of assemblies could be loaded into separate AppDomains, and communication could occur between objects as if they were one shared type. For example, if a user has a program that uses dependency A, and wants to load a plugin that has classes that inherit from older versions of dependency A, AppDomains would still be able to send and receive information between the plugin and the main program. If we enforce shared assemblies with this API, we lose the ability to allow for things like this.
+
+## Proposed Solution
+
+### Goals
+This proposal is for an API that can effectively recreate the object creation capabilities in AppDomain. The goals of this design are:
+* To allow for people are able to create isolated objects without the direct use of reflection. 
+* To preserve AppDomains ease of use when creating isolated objects.
+* To continue support for ALC unloads by destroying instantiated object when their ALC is unloaded, as well as other ALC features.
+* To allow for plugin compatibility to work even with minor versioning changes being made to the application.
+* Attempt to design around allowing inter-process communication between the user's process and the process where the object instance will be made.
+
+One thing that's especially interesting is use of this design to not only allow for communication between ALCs, but expanding the design to work for inter-process communication, which is why the 5th bullet point is there.
+
+### Design
+```C#
+interface SampleInterface {
+	public void SampleMethod();
+	...
+}
+class SampleType : SampleInterface {
+	...
+	public void SampleMethod(){
+		...
+	}
+	...
+}
+
+...
+
+
+public static void Main(){
+	AssemblyLoadContext alc = new AssemblyLoadContext(assemblyPath, true);
+	ProxyBuilder p = new ProxyBuilder(typeof(SampleInterface), alc, assemblyPath);
+	SampleInterface testObject = (SampleInterface)p.CreateObject("SampleType");
+	testObject.SampleMethod();
+	...
+
+}
+```
+The API surface allows for developers to create the ProxyBuilder object, and then create an instance of the proxy type, which works exactly like a normal instance of the target type, with the limitation that the target type must inherit from an interface that can be used to build the proxy object. This does limit the use of the object slightly, as use of the target object is limited to methods that exist in the interface.
```

Comment:
It would if we had shared assemblies, but it causes more problems if the loaded ALC isn't sharing the interface with the current context. If we just made the object in the ALC and casted it, it would cause an InvalidCastException (Basically the error picture below). The plan for the API would be to make it seem like it's a simple instantiation and cast, even if there are no shared assemblies.

Reply:
Your sample code does not make that clear.
The document should explain it.
+ Why do we need a proxy here?
+ What value does this add?